Ticket #695: new style check : explicit declaration of ctor

JakubMelka opened the pull request:
Hello,
My name is Jakub Melka and I want to help with the CppCheck development. I tried to create a patch for ticket #695 - new check for explicit constructors. I hope it will be helpful.

Best regards,
Jakub Melka

amai2012 commented:
One of the integration builds failed: https://travis-ci.org/danmar/cppcheck/jobs/53479735#L4537
That new check causes new warnings about cppcheck's own code which need to be addressed or suppressed.

danmar commented:
Thanks!

Your new checker code looks really good.

hmm.. I am not very interested in the performance problem here. I am more interested about detecting bugs caused by this.

danmar commented:
> hmm.. I am not very interested in the performance problem here. I am more interested about detecting bugs caused by this.

I am a skeptic about this checker right now based on the results.

The checker warned about several classes in Cppcheck. There were no bugs. So this checker can be noisy. I believe the fixes in valueflow makes the code more explicit so those fixes were not bad.

I understand your goal was to detect non-explicit constructors and I think you do that well.

How about writing a warning only if the constructor is used in a suspicious way also?

amaigil2 commented:
I  doubt cppcheck might claim to detect the suspicious usage. You would have to see all the code (not omit some headers which is allowed for cppcheck usage) and make a compiler quality analysis of  the required types. It is exactly the nature of this problem that the none explicit ctor gets used where it's difficult to spot.
So IMHO the only chance for cppcheck to implement this check (which is also present in some other code checkers or compilers) is complain about the declaration itself.

Indeed that is a real error but just a warning about a possible bug due to code which is open for mistakes.

danmar commented:
> I doubt cppcheck might claim to detect the suspicious usage. You would have to see all the code (not omit some headers which is allowed for cppcheck usage) and make a compiler quality analysis of the required types.

I'd still prefer this approach. however maybe it's acceptable to just warn about the declaration...

> So IMHO the only chance for cppcheck to implement this check (which is also present in some other code checkers or compilers) is complain about the declaration itself.

yes.. I guess it's ok since it is a pretty common recommendation.

I still think "performance" severity is wrong. Either "style" or "warning" is better. Hmm.. usually safety-related messages should be "warning". But this is probably quite unlikely to find actual bugs.

JakubMelka commented:
Hello,
In our company, where I work, I am a member of large team, where we are programming very huge program consisting of millions of lines of C++ code and many dll libraries. We are using CppCheck to check the C++ code. We use keyword explicit for constructors as much as possible, because we do not want unwanted type conversions (use of implicit constructor call is in our case mostly an error). I think to detect, if the constructor is used in a suspicious way, is very hard, because as amailgil2 has written, we need compiler logic to detect that. For example:

I have a class ABC in one dll, which is used in 15 other dll libraries. I want to check, if the constructor of the ABC class is not implicitly used also in the 15 other dlls. I have source code of them, but it will be hard to detect, because I must scan code of the 15 other dlls. I think it may be also a performance problem and I have to scan also include list of the every *.cpp and *.h file of these dlls, if the class ABC is included here. As I am beginner with contributing to CppCheck, I do not know how to do that. Any ideas? And what is a suspicious constructor call?

My recommendation is to simply complain about declaration itself, as it is useful, because someone can write a library with public interface, which contains non-explicit constructors and a programmer of this library can use it wrongly with implicit conversion.

I setted the "performance" severity, because it has been written in the ticket to do so. I can change that, please let me know, which severity will be better. I can change that (and I will also fix VS 2010 code).

danmar commented:
> My recommendation is to simply complain about declaration itself

ok I can allow this.

> I setted the "performance" severity, because it has been written in the ticket to do so. I can change that, please let me know, which severity will be better.

Yes of course. You did the right thing.. but the ticket was wrong.

I think we should use "style" since this is about type safety.

It would feel better for me if the message somehow says why it warns. Preferably it should be clear that there is not necessarily a bug but this message is only about type safety. Maybe:

Short message:

    Class 'X' has a constructor with 1 argument that is not explicit

Verbose message:

    Class 'X' has a constructor with 1 argument that is not explicit. Such constructors should in general be explicit for type safety reasons. Using the explicit keyword in the constructor means some mistakes when using the class can be avoided.

If you have better suggestions.. feel free to write it.

JakubMelka commented:
Hello, I have commited the desired changes:
1) Fix Visual Studio 2010 issue
2) Switched to "style" severity
3) Changed messages and updated unit tests
